File: LeetCode/wiggle_sort2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void wiggleSort(vector<int>& nums) {
        vector<int> sorted(nums);
        sort(sorted.begin(), sorted.end());
        output(sorted);
        // for(int i=0; i<nums.size()/2; i++){
        //     nums[i*2] = sorted[i];
        //     nums[i*2 + 1] = sorted[nums.size()/2 + nums.size()%2 + i];
        // }
        for(int i=nums.size()-1, j=0, k=(nums.size() - 1)/2 + 1; i>=0; i--){
            nums[i] = sorted[(i%2 ? k++: j++)];
        }
        // if(nums.size()%2){
        //     nums[nums.size() - 1] = sorted[nums.size()/2];
        //     cout << "nums[" << nums.size() - 1 << "]" << endl;
        //     cout << "sorted[" << nums.size()/2 + 1 << "] = " << sorted[nums.size()/2 + 1] << endl;
        // }
        output(nums);
    }
private:
    void output(vector<int>& input){
        for(int num: input){
            cout << num << " ";
        }
        cout << endl;
    }
};

int main(void){
    Solution sol;
    vector<int> input = {1, 1, 2, 1, 2, 2, 1};
    sol.wiggleSort(input);
    return 0;
}

File: LeetCode/wiggle_sort_2.py
class Solution(object):
    def wiggleSort(self, nums):
        tmp = sorted(nums)
        l = 0
        h = (len(nums) - 1) // 2 + 1
        for i in range(len(nums) - 1, -1, -1):
            if i % 2:
                nums[i] = tmp[h]
                h += 1
            else:
                nums[i] = tmp[l]
                l += 1
        
        return

if __name__ == "__main__":
    nums = [1, 5, 1, 1, 6, 4, 1]
    sol = Solution()
    sol.wiggleSort(nums)
    print(nums)